### PR TITLE
core: add validation if resources passed during cli exists

### PR DESCRIPTION
**Pull request by subhamkrai**
Adding validation method which checks if resources
passed during cli exists.
for example:
sourceStorageClass, destinationStorageClass, etc.

Signed-off-by: subhamkrai <srai@redhat.com>

### CODE REVIEW

**Comment by subhamkrai** [comment]
CC @travisn

File: pkg/migration/migration.go [hunk 1]
```diff
@@ -15,18 +15,18 @@ const (
...
 	// Create Kubernetes Client
 	logger.DefaultLog("Create Kubernetes Client")
 	client, err := k8sutil.NewClient(kubeConfig)
 	if err != nil {
 		return fmt.Errorf("failed to create kubernetes client: %v", err)
 	}
 
+	err = validateResources(client, sourceStorageClass, destinationStorageClass, rookNamespace, cephClusterNamespace)
+	if err != nil {
+		return fmt.Errorf("check if resources passed as argument exists: %v ", err)
```
Comment (Madhu-1) [inline pkg/migration/migration.go:27]: return fmt.Errorf("resource validation failed: %w", err)

File: pkg/migration/validate_resource.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+// validateResources checks if required areguments exists
+func validateResources(client *k8s.Clientset, sourceSC, destinationSC, rookNS, cephClusterNS string) error {
+	getOpt := v1.GetOptions{}
+	ctx := context.TODO()
+
+	_, err := client.StorageV1().StorageClasses().Get(ctx, destinationSC, getOpt)
+	if err != nil {
+		return fmt.Errorf("Destination StorageClass name %s doesn't exists. %v ", destinationSC, err)
+	}
```
Comment (Madhu-1) [inline pkg/migration/validate_resource.go:35]:
```suggestion
_, err := client.StorageV1().StorageClasses().Get(ctx, sourceSC, getOpt)
if err != nil {
if error is not found{
			return fmt.Errorf("soure StorageClass %v doesn't exists. %v ", sourceSC, err)
		}
// return failed to get storageclass
}
```

File: pkg/migration/validate_resource.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+	if sourceSC != "" {
+		_, err := client.StorageV1().StorageClasses().Get(ctx, sourceSC, getOpt)
```
Comment (Madhu-1) [inline pkg/migration/validate_resource.go:38]: new err variable declaration is not required

File: pkg/migration/validate_resource.go (hunk 2 above)
Comment (Madhu-1) [reply pkg/migration/validate_resource.go:35]: implement same check for other resources also

**Review by Madhu-1 — CHANGES_REQUESTED** [review]

File: pkg/migration/migration.go (hunk 1 above)
Comment (subhamkrai) [reply pkg/migration/migration.go]: updated

File: pkg/migration/validate_resource.go (hunk 2 above)
Comment (subhamkrai) [reply pkg/migration/validate_resource.go:35]: what other resources? It is checking if all the CLI arguments that are being passed exist.

Comment (Madhu-1) [reply pkg/migration/validate_resource.go:35]: i meant to do same error not found check for sourceSC,rookNS and cephclusterNS

Comment (subhamkrai) [reply pkg/migration/validate_resource.go:35]: maybe I'm missing something but check for `sourceSC,rookNS and cephclusterNS` is already there. They are just inside if just to check if they are not empty.

Comment (Madhu-1) [reply pkg/migration/validate_resource.go:35]: what i mean to say is add not found check to those. currently you are blindly returning not found error in case of any error

Comment (subhamkrai) [reply pkg/migration/validate_resource.go]: ahh got it. Added. Thanks

File: pkg/migration/validate_resource.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+// validateResources checks if required areguments exists
+func validateResources(client *k8s.Clientset, sourceSC, destinationSC, rookNS, cephClusterNS string) error {
+	getOpt := v1.GetOptions{}
+	ctx := context.TODO()
+
+	_, err := client.StorageV1().StorageClasses().Get(ctx, destinationSC, getOpt)
+	if err != nil && kerrors.IsNotFound(err) {
+		return fmt.Errorf("Destination StorageClass name %s doesn't exists. %v ", destinationSC, err)
+	}
```
Comment (Madhu-1) [inline pkg/migration/validate_resource.go:36]: don't you need to return the error other than not found? :thinking:

Comment (subhamkrai) [reply pkg/migration/validate_resource.go:39]: right. added valid check and return

File: pkg/migration/validate_resource.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+func validateResources(client *k8s.Clientset, sourceSC, destinationSC, rookNS, cephClusterNS string) error {
+	getOpt := v1.GetOptions{}
+	ctx := context.TODO()
+
+	_, err := client.StorageV1().StorageClasses().Get(ctx, destinationSC, getOpt)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			return fmt.Errorf("Destination StorageClass %s doesn't exists. %w ", destinationSC, err)
+		}
+		return fmt.Errorf("Failed to get Destination StorageClass %s. %w ", destinationSC, err)
```
Comment (Madhu-1) [inline pkg/migration/validate_resource.go:38]: error messages should start with a small case and remove extra space of `%w`

Comment (subhamkrai) [reply pkg/migration/validate_resource.go]: done

File: pkg/migration/validate_resource.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+		}
+	}
+
+	if rookNS != "" {
+		_, err = client.CoreV1().Namespaces().Get(ctx, rookNS, getOpt)
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				return fmt.Errorf("rook namespace %s doesn't exists. %w", rookNS, err)
+			}
+			return fmt.Errorf("failed to get Rook namespace  %s. %w", rookNS, err)
```
Comment (Madhu-1) [inline pkg/migration/validate_resource.go:57]: remove extra space after `Rook namespace`

**Review by Madhu-1 — APPROVED** [review]
LGTM

File: pkg/migration/validate_resource.go (hunk 6 above)
Comment (subhamkrai) [reply pkg/migration/validate_resource.go]: added

File: pkg/migration/validate_resource.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+// validateResources checks if required areguments exists
+func validateResources(client *k8s.Clientset, sourceSC, destinationSC, rookNS, cephClusterNS string) error {
+	getOpt := v1.GetOptions{}
+	ctx := context.TODO()
+
+	_, err := client.StorageV1().StorageClasses().Get(ctx, destinationSC, getOpt)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			return fmt.Errorf("destination storageClass %s doesn't exists. %w", destinationSC, err)
```
Comment (travisn) [inline pkg/migration/validate_resource.go:36]: All of the messages in this function should be "exist" instead of "exists"
```suggestion
			return fmt.Errorf("destination storageClass %s doesn't exist. %w", destinationSC, err)
```

File: pkg/migration/migration.go [hunk 8]
```diff
@@ -15,18 +15,18 @@ const (
...
 	// Create Kubernetes Client
 	logger.DefaultLog("Create Kubernetes Client")
 	client, err := k8sutil.NewClient(kubeConfig)
 	if err != nil {
 		return fmt.Errorf("failed to create kubernetes client: %v", err)
 	}
 
+	err = validateResources(client, sourceStorageClass, destinationStorageClass, rookNamespace, cephClusterNamespace)
+	if err != nil {
+		return fmt.Errorf("resource validation failed: %w", err)
```
Comment (travisn) [inline pkg/migration/migration.go:27]:
```suggestion
		return errors.Wrap(err, "resource validation failed")
```

**Review by travisn — COMMENTED** [review]
just a couple nits

File: pkg/migration/validate_resource.go (hunk 7 above)
Comment (subhamkrai) [reply pkg/migration/validate_resource.go]: done.

File: cmd/cmd.go [hunk 9]
```diff
@@ -66,7 +66,7 @@ func init() {
 	// will be global for your application.
 
 	rootCmd.PersistentFlags().StringVar(&kubeConfig, "kubeconfig", "", "kubernetes config path")
-	rootCmd.PersistentFlags().StringVar(&sourceStorageClass, "sourcestoraageclass", "", "source storageclass from which all PVC need to be migrated")
+	rootCmd.PersistentFlags().StringVar(&sourceStorageClass, "sourcestorageclass", "", "source storageclass from which all PVC need to be migrated")
```
Comment (travisn) [inline cmd/cmd.go:69]:
```suggestion
	rootCmd.PersistentFlags().StringVar(&sourceStorageClass, "source-storageclass", "", "source storageclass from which all PVC need to be migrated")
```

File: cmd/cmd.go [hunk 10]
```diff
@@ -66,7 +66,7 @@ func init() {
...
 	rootCmd.PersistentFlags().StringVar(&destinationStorageClass, "destinationstorageclass", "", "destination storageclass (CSI storageclass) to which all PVC need to be migrated")
```
Comment (travisn) [inline cmd/cmd.go:70]: Maybe these should be a separate PR, but seems good for consistency
```suggestion
	rootCmd.PersistentFlags().StringVar(&destinationStorageClass, "destination-storageclass", "", "destination storageclass (CSI storageclass) to which all PVC need to be migrated")
```

File: pkg/migration/validate_resource.go [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+// validateResources checks if required areguments exists
+func validateResources(client *k8s.Clientset, sourceSC, destinationSC, rookNS, cephClusterNS string) error {
+	getOpt := v1.GetOptions{}
+	ctx := context.TODO()
+
+	_, err := client.StorageV1().StorageClasses().Get(ctx, destinationSC, getOpt)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			return fmt.Errorf("destination storageClass %s doesn't exist. %w", destinationSC, err)
```
Comment (travisn) [inline pkg/migration/validate_resource.go:36]: Do we need a special error message if it isn't found? I would think the err returned by the k8s api would give a message anyway about it not being found. But if our message is more clear, sounds good to keep it since this is the most common error we expect.

File: cmd/cmd.go (hunk 9 above)
Comment (subhamkrai) [reply cmd/cmd.go:69]: already part of #27 as diff commit but can make separate PR

File: pkg/migration/validate_resource.go (hunk 11 above)
Comment (subhamkrai) [reply pkg/migration/validate_resource.go:36]: okay, sounds good to have the err message.

**Review by Yuggupta27 — APPROVED** [review]
lgtm
